Provide ddl support for presto-on-spark library

sambekar15 opened the pull request:
Test plan - 
Tested DROP and GRANT statements via spark submit using presto-spark library .
`  spark-submit \
--master spark://spark-master:7077 \
--executor-cores 8 \
--conf spark.task.cpus=8 \
--class com.facebook.presto.spark.launcher.PrestoSparkLauncher \
  presto-spark-launcher-0.262-SNAPSHOT.jar \
--package presto-spark-package-0.262-SNAPSHOT.tar.gz \
--config config.properties \
--catalogs catalogs \
--catalog hive \
--schema default \
--file drop.sql`

drop.sql
`DROP table if exists schema1.table1`

grant.sql
`GRANT SELECT ON schema1.table1 TO ROLE PUBLIC
`
Added tests for DDL Support in TestPrestoQuerySparkRunner.

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines. Don't forget to follow our [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution) for any code copied from other projects.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

Presto-On-Spark Changes
* Provide DDL support 
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

aweisberg commented on presto-spark-base/src/test/java/com/facebook/presto/spark/PrestoSparkQueryRunner.java:
```diff
@@ -293,7 +294,10 @@ public PrestoSparkQueryRunner(String defaultCatalog, Map<String, String> additio
         connectorManager.createConnection("hive", "hive", ImmutableMap.of(
-                "hive.experimental-optimized-partition-update-serialization-enabled", "true"));
+                "hive.experimental-optimized-partition-update-serialization-enabled", "true",
+                "hive.allow-drop-table", "true", "hive.allow-rename-table", "true",
```
I am trying to understand why this was necessary to test DDL when we don't use this approach to configure security in any other tests. It seems like HiveQueryRunner generally uses "sql-standard".

That sent me down the rabbit hole of trying to understand what security OSS PoS uses and it seems like it would get the default "legacy".

arhimondr replied:
nit: one property per line

sambekar15 replied:
I am getting access denied exception if I don't set these properties 
`com.facebook.presto.spi.security.AccessDeniedException: Access Denied: Cannot drop table XXXX
	at com.facebook.presto.spi.security.AccessDeniedException.denyDropTable(AccessDeniedException.java:116)
	at com.facebook.presto.spi.security.AccessDeniedException.denyDropTable(AccessDeniedException.java:111)
	at com.facebook.presto.hive.security.LegacyAccessControl.checkCanDropTable(LegacyAccessControl.java:100)
	at com.facebook.presto.hive.security.SystemTableAwareAccessControl.checkCanDropTable(SystemTableAwareAccessControl.java:88)
	at com.facebook.presto.security.AccessControlManager.lambda$checkCanDropTable$20(AccessControlManager.java:304)
	at com.facebook.presto.security.AccessControlManager.authorizationCheck(AccessControlManager.java:738)
	at com.facebook.presto.security.AccessControlManager.checkCanDropTable(AccessControlManager.java:304)
	at com.facebook.presto.execution.DropTableTask.execute(DropTableTask.java:67)
`

aweisberg commented on presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java:
```diff
@@ -397,51 +407,72 @@ public IPrestoSparkQueryExecution create(
+                        StatementUtils.isTransactionControlStatement(preparedQuery.getStatement()),
+                        transactionManager,
+                        accessControl,
+                        directExecutor(),
```
Why direct executor? Usually when an executor service is requested there is an expectation it's going to provide a dedicated thread.

aweisberg commented on presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDataDefinitionExecution.java:
```diff
@@ -0,0 +1,110 @@
+    public List<List<Object>> execute()
+    {
+        log.info("Executing DDL statemt %s", statement.toString());
+
```
Why no transition to running and the try catch we see in `DataDefinitionExecution`?

aweisberg commented on presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDataDefinitionExecution.java:
```diff
@@ -0,0 +1,110 @@
+            @Override
+            public void onSuccess(@Nullable Object result)
+            {
+                getFutureValue(transactionManager.asyncCommit(getTransactionInfo(queryStateMachine.getSession(), transactionManager).getTransactionId()));
```
This also differs. It looks like this is bypassing issues with `QueryStateMachine`. Would it make sense to integrate fully with `QueryStateMachine` so it works for both PoS and Classic?

aweisberg reviewed (changes requested):
First round of changes/questions. Thank you for the contribution!

arhimondr commented on presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDataDefinitionExecution.java:
```diff
@@ -0,0 +1,110 @@
+
+        ListenableFuture<?> future = task.execute(statement, transactionManager, metadata, accessControl, queryStateMachine, Collections.emptyList());
+
+        Futures.addCallback(future, new FutureCallback<Object>()
```
We should wait on this feature before returning. Error and transaction handling can be done synchronously then as well.

arhimondr commented on presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDataDefinitionExecution.java:
```diff
@@ -0,0 +1,110 @@
+
+    private void fail(Throwable cause)
+    {
+        queryStateMachine.transitionToFailed(cause);
```
Should the transaction be reverted upon failure?

arhimondr commented on presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDataDefinitionExecution.java:
```diff
@@ -0,0 +1,110 @@
+    private void fail(Throwable cause)
+    {
+        queryStateMachine.transitionToFailed(cause);
+        queryStateMachine.updateQueryInfo(Optional.empty());
```
For PoS query state machine should be unnecesary. Have we considered refactoring tasks not to require the state machine?

sambekar15 replied:
yeah correct  QueryStateMachine is not necessary for PoS. So initially I started refactoring tasks to remove dependency on QueryStateMachine. I was able to successfully refactor Drop,Grant,Rename,Create tasks but the transactional tasks - SetSessionTask,ResetSessionTask,RollbackTask,CommitTask,StartTransactionTask,UseTask,PrepareTask had  dependency on other methods of QuerystateMachine. It involved refactoring all tasks and i was afraid that it may cause regression but we have enough tests to catch these issues. I agree this may be better approach (to reduce dependency on QueryStateMachine). Let me know your thoughts and I can follow that approach.

arhimondr commented on presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDataDefinitionExecution.java:
```diff
@@ -0,0 +1,110 @@
+    {
+        log.info("Executing DDL statemt %s", statement.toString());
+
+        ListenableFuture<?> future = task.execute(statement, transactionManager, metadata, accessControl, queryStateMachine, Collections.emptyList());
```
nit: static import `emptyList`

aweisberg commented:
@sambekar15  is this ready for review? You can click the request re-request review in the list of reviewers. Also looking forward to the multi-statement enhancement :-)

sambekar15 commented:
> @sambekar15 is this ready for review? You can click the request re-request review in the list of reviewers. Also looking forward to the multi-statement enhancement :-)

@aweisberg @arhimondr yes, its ready for review now. 
Summary of my changes -
1) Created a helper class to bind statements to `DataDefinitionTasks`. Categorized the DataDefinition tasks into transaction/session and regular ddl tasks. Coordinator module binds all DataDefinition tasks to statements. Spark module binds only ddl tasks to statements , so transaction and session statements is not supported for PoS.
2) Added a new interface `SessionTransactionControlTask` extending `DataDefinitionTask`. This Interface is specifically for tasks needing QueryStateMachine for transaction and session management. So SetSesstion, ResetSession, Prepare,Deallocate, Use, Commit, Transaction tasks would implement this interface. 
3) Grant tests  added in PoS were failing initially. Noticed `FileHiveMetastore` accessing `PermissionMetadata` was using hardcoded user admin. Refactored the class to take user as parameter. Tests are passing after change. 

I am also trying to remove dependency of QueryStateMachine from these `SessionTransactionControlTask` and instead use shared Session/transaction but it requires more refactoring. I can take this up when providing multi-statement enhancement for PoS , so that these statements can run in PoS as well.

I see a couple of assertion tests (in presto-product-tests, presto-raptor) failing in PR build . These tests passed on my local though. I'll check.

Please let me know your feedback when you get chance.

aweisberg reviewed (changes requested):
This looks like it is in a pretty reasonable place, but I have questions about the interface change to `DataDefinitionTask` and `SessionTransactionControlTask`. Seems like that might be better using `Optional` for `QueryStateMachine`, WDYT?

sambekar15 commented:
> This looks like it is in a pretty reasonable place, but I have questions about the interface change to `DataDefinitionTask` and `SessionTransactionControlTask`. Seems like that might be better using `Optional` for `QueryStateMachine`, WDYT?

Thanks for your feedback. Yeah i did consider using `DataDefinitionTask`  to use `Optional` for `QueryStateMachine` but then went with `SessionTransactionControlTask` as `Optional` can allow to pass null values.  yeah , changing execute method in `DataDefinitionTask`  to use `Optional` for `QueryStateMachine` will make things simpler and will resolve rest of your comments.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/SessionTransactionControlTask.java:
```diff
@@ -0,0 +1,37 @@
+{
+    ListenableFuture<?> execute(T statement, TransactionManager transactionManager, Metadata metadata,
+                                AccessControl accessControl, Session session, List<Expression> parameters,
+                                WarningCollector warningCollector, QueryStateMachine stateMachine);
```
I would probably recommend having two separate interfaces: `DataDefinitionTask` and `SessionControlTask`. The first one (`DataDefinitionTask`) would have a method that takes a `Session` and a `WarningCollector` collector and don't require a `QueryStateMachine` that is not available in `Presto on Spark`. The second interface would take the `QueryStateMachine` but not `Session` and `WarningCollector` as those can be obtained from the `QueryStateMachine`. Than with a little of `Guice` magic these two interfaces can be bounded into two separate providers and selected based on the statement type.

aweisberg replied:
OK we can go with Andrii's approach since he is the one committing it.

sambekar15 replied:
@aweisberg and @arhimondr  I have pushed these changes. Please review when you get chance . Thanks.

aweisberg commented:
Can you start the CLA workflow? It's tricky with multiple steps so might as well get started. Reviewing now.

aweisberg commented on presto-main/src/main/java/com/facebook/presto/execution/SessionDefinitionExecution.java:
```diff
@@ -0,0 +1,160 @@
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.util.Objects.requireNonNull;
+
+public class SessionDefinitionExecution<T extends Statement>
```
This class largely copies the contents of `DataDefinitionExecution` and it also isn't really a `DataDefinitionExecution`? 

Really it seems like there should be a common base class for both of them and they should only contain what is specific to them?

aweisberg commented on presto-main/src/main/java/com/facebook/presto/execution/SessionTransactionControlTask.java:
```diff
@@ -0,0 +1,40 @@
+                                AccessControl accessControl, QueryStateMachine stateMachine,
+                                List<Expression> parameters);
+
+    default boolean isTransactionControl()
```
This is unused now?

aweisberg reviewed (changes requested):
I think this is closer, but some things still have relationships that duplicate or implement extra things.

There are some unused functions in QueryStateMachine now.

aweisberg reviewed (approved):
LGTM, thank you!

arhimondr commented on presto-main/src/main/java/com/facebook/presto/execution/UseTask.java:
```diff
@@ -24,15 +24,18 @@
-        implements DataDefinitionTask<Use>
+        implements SessionTransactionControlTask<Use>
 {
+    private final AtomicReference<String> setCatalog = new AtomicReference<>();
```
Is it needed?

aweisberg commented:
Looks like you still need to sign the CLA and there is one test failure. Andrii also had a comment :-)

sambekar15 commented:
> Looks like you still need to sign the CLA and there is one test failure. Andrii also had a comment :-)

@aweisberg @arhimondr I have rebased and squashed my commits. CLA error is fixed. All tests have passed . I have also addressed the comment for unused variable.
